### ISSUE TITLE
Server freeze/crash

### ISSUE THREAD

**Issue by iLemon** [IntellectualSites/FastAsyncWorldEdit#418]
# Bug report for FastAsyncWorldEdit 1.15.2

**[REQUIRED] FastAsyncWorldEdit Configuration Files:** 

 https://athion.net/ISPaster/paste/view/0dfde2bb97d842f584817c57e44d7018

**[REQUIRED] FastAsyncWorldEdit Version Number:** 

```
> fawe version
[15:25:02 INFO]: FastAsyncWorldEdit-177 created by Empire92
[15:25:02 INFO]: FAWE version:
 - Date Apr 23, 2020, 12:00:00 AM
 - Commit fd8c1731
 - Build 177
 - Platform Bukkit
[15:25:02 INFO]: Wiki: https://github.com/IntellectualSites/FastAsyncWorldEdit-1.13/wiki
[15:25:02 INFO]: ------------------ Platforms ------------------
* Bukkit-Official(1.15-177;273e8cf)

[15:25:02 INFO]: ---------------- Capabilities ----------------
GAME_HOOKS: Bukkit-Official
CONFIGURATION: Bukkit-Official
USER_COMMANDS: Bukkit-Official
PERMISSIONS: Bukkit-Official
WORLDEDIT_CUI: Bukkit-Official
WORLD_EDITING: Bukkit-Official
```

**[REQUIRED] Spigot/Paper Version Number:** 

```
> ver
[15:24:45 INFO]: This server is running Citrus version git-Citrus-f8d15d7 (MC: 1.15.2) (Implementing API version 1.15.2-R0.1-SNAPSHOT)
[15:24:45 INFO]: Previous version: git-Citrus-21b00e5 (MC: 1.15.2)
[15:24:45 INFO]: Unknown version
```

**[REQUIRED] Description of the problem:**

https://gist.githubusercontent.com/iLemon/8095eae14c9329ff804e566a781630c4/raw/4fbbcf73a124bb24b0e6ca55133b6abaf3bce73b/gistfile1.txt

**Comment by MattBDev** [comment]
@iLemon How many people were on the server before the crash and can you describe what people were doing if possible?

**Comment by iLemon** [comment]
There were about 40 players online.

It's a creative plots server. Players have access to restricted worldedit with the following permissions:
```
[16:05:48 INFO]: > fawe.worldguard (true)
[16:05:48 INFO]: > fawe.worldguardflag (true)
[16:05:48 INFO]: > worldedit.analysis.count (true)
[16:05:48 INFO]: > worldedit.analysis.size (true)
[16:05:48 INFO]: > worldedit.biome.list (true)
[16:05:48 INFO]: > worldedit.biome.set (true)
[16:05:48 INFO]: > worldedit.clipboard.copy (true)
[16:05:48 INFO]: > worldedit.clipboard.cut (true)
[16:05:48 INFO]: > worldedit.clipboard.flip (true)
[16:05:48 INFO]: > worldedit.clipboard.paste (true)
[16:05:48 INFO]: > worldedit.clipboard.rotate (true)
[16:05:48 INFO]: > worldedit.fill (true)
[16:05:48 INFO]: > worldedit.fixwater (true)
[16:05:48 INFO]: > worldedit.history.redo (true)
[16:05:48 INFO]: > worldedit.history.undo (true)
[16:05:54 INFO]: > worldedit.navigation.ascend (true)
[16:05:54 INFO]: > worldedit.navigation.descend (true)
[16:05:54 INFO]: > worldedit.navigation.unstuck (true)
[16:05:54 INFO]: > worldedit.region.centre (true)
[16:05:54 INFO]: > worldedit.region.faces (true)
[16:05:54 INFO]: > worldedit.region.forest (true)
[16:05:54 INFO]: > worldedit.region.move (true)
[16:05:54 INFO]: > worldedit.region.naturalize (true)
[16:05:54 INFO]: > worldedit.region.overlay (true)
[16:05:54 INFO]: > worldedit.region.replace (true)
[16:05:54 INFO]: > worldedit.region.set (true)
[16:05:54 INFO]: > worldedit.region.stack (true)
[16:05:54 INFO]: > worldedit.region.walls (true)
[16:05:54 INFO]: > worldedit.removeabove (true)
[16:05:54 INFO]: > worldedit.removebelow (true)
[16:05:54 INFO]: > worldedit.removenear (true)
[16:05:54 INFO]: > worldedit.replacenear (true)
[16:05:54 INFO]: > worldedit.selection.pos (true)
[16:05:54 INFO]: > worldedit.superpickaxe (true)
[16:05:55 INFO]: > worldedit.wand (true)
```

**Comment by iLemon** [comment]
If it helps, these are all the worldedit related commands that were run between server startup and this 'crash': https://gist.githubusercontent.com/iLemon/079e7074a9b4d7cb5459291ab34a4279/raw/35f4686df928d60c7decb7da5a9ebb80c866866d/gistfile1.txt

The server shuts down after this error.

**Comment by dordsor21** [comment]
Given the fact that further crashes haven't had the same error, I'm not really sure it's necessarily FAWE causing the issue. No other users have reported the issue, and it's inconsistent with the WE commands being used beforehand.

Keeping open just in case.

**Comment by stevekr3w** [comment]
Same here... 

> -- Head --
Thread: Server thread
Stacktrace:
	at com.boydti.fawe.bukkit.adapter.DelegateLock.isHeldByCurrentThread(DelegateLock.java:81)
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.a(SourceFile:40)
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.onResize(SourceFile:92)
	at net.minecraft.server.v1_15_R1.DataPaletteLinear.a(SourceFile:44)
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.a(SourceFile:122)
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.setBlock(SourceFile:112)
	at net.minecraft.server.v1_15_R1.ChunkSection.setType(SourceFile:60)
	at net.minecraft.server.v1_15_R1.ChunkSection.setType(SourceFile:54)
	at net.minecraft.server.v1_15_R1.Chunk.setType(Chunk.java:268)
	at net.minecraft.server.v1_15_R1.World.setTypeAndData(World.java:224)
	at net.minecraft.server.v1_15_R1.BlockLeaves.tick(BlockLeaves.java:41)
	at net.minecraft.server.v1_15_R1.IBlockData.a(SourceFile:263)
	at net.minecraft.server.v1_15_R1.WorldServer.b(WorldServer.java:599)

> -- Block being ticked --

Details:
	Block location: World: (1537,81,-288), Chunk: (at 1,5,0 in 96,-18; contains blocks 1536,0,-288 to 1551,255,-273), Region: (3,-1; contains chunks 96,-32 to 127,-1, blocks 1536,0,-512 to 2047,255,-1)
Stacktrace:
	at net.minecraft.server.v1_15_R1.TickListServer.b(TickListServer.java:82)
	at net.minecraft.server.v1_15_R1.WorldServer.doTick(WorldServer.java:312)



**SERVERLOG:**

> [21:10:38] [Server thread/INFO]: ********** issued server command: /day
[21:10:44] [Server thread/INFO]: ********** issued server command: //paste
[21:11:32] [Server thread/INFO]: ********** issued server command: //copy
[21:11:35] [Server thread/INFO]: ********** issued server command: //undo
[21:11:42] [Server thread/INFO]: ********** issued server command: //undo
[21:11:55] [Server thread/INFO]: ********** issued server command: //paste
[21:15:38] [Server thread/INFO]: ********** issued server command: //set 0
[21:17:26] [Server thread/INFO]: ********** issued server command: /day
[21:23:23] [Server thread/INFO]: ********** issued server command: /gm 3
[21:23:25] [Server thread/INFO]: ********** issued server command: /day
[21:23:28] [Server thread/INFO]: ********** issued server command: /tp **********
[21:25:23] [Server thread/INFO]: ********** issued server command: //replace 1 2
[21:25:38] [Server thread/INFO]: ********** issued server command: //replace 1 2
[21:26:00] [Server thread/INFO]: ********** issued server command: /gm 1
[21:26:16] [Server thread/INFO]: ********** issued server command: /gm 1 *********
[21:27:56] [Server thread/INFO]: ********** issued server command: //replace 3 2
[21:31:11] [Server thread/INFO]: ********** issued server command: /day
[21:34:32] [Server thread/ERROR]: Encountered an unexpected exception
net.minecraft.server.v1_15_R1.ReportedException: Exception while ticking
	at net.minecraft.server.v1_15_R1.MinecraftServer.b(MinecraftServer.java:1095) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.DedicatedServer.b(DedicatedServer.java:406) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.MinecraftServer.a(MinecraftServer.java:984) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.MinecraftServer.run(MinecraftServer.java:824) [spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_201]
Caused by: java.lang.UnsupportedOperationException
	at com.boydti.fawe.bukkit.adapter.DelegateLock.isHeldByCurrentThread(DelegateLock.java:81) ~[?:?]
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.a(SourceFile:40) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.onResize(SourceFile:92) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.DataPaletteLinear.a(SourceFile:44) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.a(SourceFile:122) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.setBlock(SourceFile:112) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.ChunkSection.setType(SourceFile:60) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.ChunkSection.setType(SourceFile:54) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.Chunk.setType(Chunk.java:268) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.World.setTypeAndData(World.java:224) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.BlockLeaves.tick(BlockLeaves.java:41) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.IBlockData.a(SourceFile:263) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.WorldServer.b(WorldServer.java:599) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.TickListServer.b(TickListServer.java:82) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.WorldServer.doTick(WorldServer.java:312) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.MinecraftServer.b(MinecraftServer.java:1082) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	... 4 more
[21:34:32] [Server thread/ERROR]: 	Cause of unexpected exception was
java.lang.UnsupportedOperationException: null
	at com.boydti.fawe.bukkit.adapter.DelegateLock.isHeldByCurrentThread(DelegateLock.java:81) ~[?:?]
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.a(SourceFile:40) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.onResize(SourceFile:92) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.DataPaletteLinear.a(SourceFile:44) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.a(SourceFile:122) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.setBlock(SourceFile:112) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.ChunkSection.setType(SourceFile:60) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.ChunkSection.setType(SourceFile:54) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.Chunk.setType(Chunk.java:268) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.World.setTypeAndData(World.java:224) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.BlockLeaves.tick(BlockLeaves.java:41) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.IBlockData.a(SourceFile:263) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.WorldServer.b(WorldServer.java:599) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.TickListServer.b(TickListServer.java:82) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.WorldServer.doTick(WorldServer.java:312) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.MinecraftServer.b(MinecraftServer.java:1082) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.DedicatedServer.b(DedicatedServer.java:406) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.MinecraftServer.a(MinecraftServer.java:984) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.MinecraftServer.run(MinecraftServer.java:824) [spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_201]
[21:34:32] [Server thread/ERROR]: This crash report has been saved to: *********/*********/./crash-reports/crash-2020-04-25_21.34.32-server.txt
[21:34:32] [Server thread/INFO]: Stopping server
[21:34:33] [Server thread/INFO]: Saving worlds
[21:34:33] [Server thread/INFO]: Saving chunks for level 'world'/minecraft:overworld
[21:34:33] [Server thread/ERROR]: Failed to save chunk 96,-18
java.lang.UnsupportedOperationException: null
	at com.boydti.fawe.bukkit.adapter.DelegateLock.isHeldByCurrentThread(DelegateLock.java:81) ~[?:?]
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.a(SourceFile:40) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.DataPaletteBlock.a(SourceFile:204) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.ChunkRegionLoader.saveChunk(ChunkRegionLoader.java:259) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.PlayerChunkMap.saveChunk(PlayerChunkMap.java:717) ~[spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174) [?:1.8.0_201]
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175) [?:1.8.0_201]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) [?:1.8.0_201]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382) [?:1.8.0_201]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) [?:1.8.0_201]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) [?:1.8.0_201]
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151) [?:1.8.0_201]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174) [?:1.8.0_201]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) [?:1.8.0_201]
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418) [?:1.8.0_201]
	at net.minecraft.server.v1_15_R1.PlayerChunkMap.save(PlayerChunkMap.java:317) [spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.ChunkProviderServer.save(ChunkProviderServer.java:302) [spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.WorldServer.save(WorldServer.java:811) [spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.MinecraftServer.saveChunks(MinecraftServer.java:655) [spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.MinecraftServer.stop(MinecraftServer.java:725) [spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.DedicatedServer.stop(DedicatedServer.java:696) [spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at net.minecraft.server.v1_15_R1.MinecraftServer.run(MinecraftServer.java:864) [spigot-1.15.2.jar:git-Spigot-e7dc2f5-a8ec1d6]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_201]
[21:34:34] [Server thread/INFO]: ThreadedAnvilChunkStorage (world): All chunks are saved
[21:34:34] [Server thread/INFO]: Saving chunks for level 'world_nether'/minecraft:the_nether
[21:34:34] [Server thread/INFO]: ThreadedAnvilChunkStorage (DIM-1): All chunks are saved
[21:34:34] [Server thread/INFO]: Saving chunks for level 'world_the_end'/minecraft:the_end
[21:34:34] [Server thread/INFO]: ThreadedAnvilChunkStorage (DIM1): All chunks are saved
[21:34:34] [Server thread/INFO]: Saving chunks for level 'games'/minecraft:games
[21:34:34] [Server thread/INFO]: ThreadedAnvilChunkStorage (games): All chunks are saved
[21:34:34] [Server thread/INFO]: ThreadedAnvilChunkStorage (world): All chunks are saved
[21:34:35] [Server thread/INFO]: ThreadedAnvilChunkStorage (DIM-1): All chunks are saved
[21:34:35] [Server thread/INFO]: ThreadedAnvilChunkStorage (DIM1): All chunks are saved
[21:34:35] [Server thread/INFO]: ThreadedAnvilChunkStorage (games): All chunks are saved

**Comment by IronApollo** [comment]
Doogie has just reported this timeout as well. I'll be changing tags to reflect this as tested, bug and priority.

**Comment by MattBDev** [comment]
@IronApollo Was it an UnsupportedOperationException?

**Comment by IronApollo** [comment]
@MattBDev No, but after further evaluation I don't believe Doogie's is the same as this issue, it seems more related to the DelegateLock issue we've seen before.

**Comment by dordsor21** [comment]
To me that looks like the server attempting to place a block at the same time a chunk is being edited. I think we need to disable and then re-enable ticking in the chunk/ChunkSection when editing

**Comment by aurorasmiles** [comment]
Wasn't this fixed in one of the later versions & paper version?

**Comment by aurorasmiles** [comment]
Assuming this was fixed in 0dee452 and https://github.com/PaperMC/Paper/commit/68994c64ec03b557bf244df2b3a9ad3810aec550
